spec/requests/provider_interface/provider_user_impersonates_candidate_spec.rb

```diff
@@ -0,0 +1,48 @@
+require 'rails_helper'
+
+RSpec.describe 'POST /provider/candidates/:id/impersonate' do
+  include CourseOptionHelpers
+
+  context 'when the user is logged in to Apply' do
+    let(:provider) { create(:provider, :with_signed_agreement) }
+    let(:provider_user) { create(:provider_user, providers: [provider], dfe_sign_in_uid: 'DFE_SIGN_IN_UID') }
+    let(:course_option) { course_option_for_provider_code(provider_code: provider.code) }
+    let(:application_choice) { create(:application_choice, status: 'awaiting_provider_decision', course_option: course_option) }
+
+    before do
+      allow(DfESignInUser).to receive(:load_from_session)
+        .and_return(
+          DfESignInUser.new(
+            email_address: provider_user.email_address,
+            dfe_sign_in_uid: provider_user.dfe_sign_in_uid,
+            first_name: provider_user.first_name,
+            last_name: provider_user.last_name,
+          ),
+      )
+    end
+
+    it 'redirects to Candidate Interface if candidate associated with user\'s providers' do
+      post provider_interface_impersonate_candidate_path(application_choice.application_form.candidate)
+
+      expect(response).to have_http_status 302
+      expect(response.body).to match(/\/candidate/)
+    end
+
+    it 'redirects back to Provider Interface if candidate is not associated with user\'s providers' do
+      unrelated_application = create(:application_choice, status: 'awaiting_provider_decision')
+
+      post provider_interface_impersonate_candidate_path(unrelated_application.application_form.candidate)
+
+      expect(response).to have_http_status 302
+      expect(response.body).to match(/\/provider/)
```

Comment:
Request spec feels like the right things here :+1:. The assertions on `provider` and `candidate` (at l.28) aren't too decisive. Is there something stronger we could match on to prove this works.

Reply:
It's tricky precisely because it's a request spec. I've added an attempt to access the application form in candidate_interface, which gives 200 only if you have an active candidate session.

Reply:
I think that's great :100: